Implements barycentric interpolation on a multi-dimensional mesh

RussTedrake opened the pull request:
This will be the backend for the dynamic programming algorithms that should follow quickly.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8023)
<!-- Reviewable:end -->

RussTedrake commented:
+@hongkai-dai for feature review, please.  (the CI failure is unrelated)

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-:-L4wsus0ZWrGolXrZ-SD:b-v0516j)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-memcheck-lsan please

jwnimmer-tri commented:
Checkpoint -- about halfway through.

---

Reviewed 1 of 4 files at r1.
Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[math/barycentric.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4xLPX3hXl5wTRT31Qy:-L4xLPX3hXl5wTRT31Qz:bgrzhq2) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L26)):*
> ```cpp
>     DRAKE_DEMAND(!input_grid_->at(i).empty());
> 
>     // Gain one interpolant for every non-singleton dimension
> ```

BTW punc

---

*[math/barycentric.h, line 48 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4x7iKe6_hxEYFLj8dC:-L4x7iKfhmOrPPS3ydgL:b-4hxsmt) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L48)):*
> ```objc
> 
>  public:
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(BarycentricMesh);
> ```

BTW Missing `#include` for drake_copyable.

---

*[math/barycentric.h, line 57 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-OmpW-amSL0T6usmI:-L5-OmpXNXgStlyGfb0m:bspehhq) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L57)):*
> ```objc
> 
>   /// Constructs the mesh.
>   explicit BarycentricMesh(std::unique_ptr<MeshGrid> input_grid);
> ```

FYI The `input_grid` could also be passed by-value (moved).  Example patch posted at https://gist.github.com/jwnimmer-tri/37b27098f2a28ca34a099bdb68426eda

Advantages of unique_ptr formulation:
- required if `MeshGrid` in the future might be turned into a class that was non-copyable;
- prevents accidentally deep copying (vs forgetting the std::move with by-value).

Advantages of by-value formulation:
- call sites can use initializer lists (i.e., minimal syntax, just the doubles);
- pydrake might be easier to wrap with the plain value.

---

*[math/barycentric.h, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4x81mMPIJHfOzcZyio:-L4x81mMPIJHfOzcZyip:b-93o217) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L59)):*
> ```objc
>   explicit BarycentricMesh(std::unique_ptr<MeshGrid> input_grid);
> 
>   /// Accessor methods.
> ```

BTW This comment only binds to the first method.  Suggest either making it `//` or else using Doxygen grouping syntax.

---

*[math/test/barycentric_test.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4xLSnHdtvgCt9HrUm7:-L4xLSnHdtvgCt9HrUm8:bh0hd81) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L29)):*
> ```cpp
>   auto grid = std::make_unique<BarycentricMesh<double>::MeshGrid>(kNumInputs);
> 
>   // Setup the input mesh by specifying the grid one dimension at at time.
> ```

BTW double `at at`

---

*[math/test/barycentric_test.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4xLVe2u8AR3GuAVlhN:-L4xLVe2u8AR3GuAVlhO:b-nvvdzb) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L50)):*
> ```cpp
> 
> // Create a mesh in 3 (input) dimensions, with the second dimension being a
> // singleton.
> ```

BTW This duplicateS the comment immediately below; this comment doesn't really appear to describe the case itself at all.

---

*[math/test/barycentric_test.cc, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4xLd2eGmd53nvUfl7L:-L4xLd2eGmd53nvUfl7M:bu0hvei) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L127)):*
> ```cpp
>   EXPECT_NEAR(value[0], 4., tol);
> 
>   // Check the middle
> ```

BTW _Meta_ punc is missing on some of the comments like this (though not all of them are sentences, this and others are).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-:-L5-SNh9wqCmghEciOPX:b-owwrjn)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
First pass done.

---

Reviewed 3 of 4 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[math/barycentric.cc, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-XClh9JluSU5O3f3K:-L5-XClh9JluSU5O3f3L:b-wf5ncg) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L51)):*
> ```cpp
> template <typename T>
> void BarycentricMesh<T>::EvalBarycentricWeights(
>     const Eigen::Ref<const VectorX<T>> input,
> ```

BTW Should we fail fast if `input` has the wrong size?

---

*[math/barycentric.cc, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-Y5uyIIHmsQ4GzJZX:-L5-Y5uyIIHmsQ4GzJZY:ba35uvg) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L56)):*
> ```cpp
>   // std::pair of fractional position [0,1] and dimension index (position first,
>   // so that std::pair's default operator< works for us).
>   std::vector<std::pair<T, int>> relative_position(num_interpolants_ - 1);
> ```

FYI If `T` is `AutoDiffScalar` I believe this will leave is `.value()` initialized to garbage.  We don't need to solve that yet, though.

---

*[math/barycentric.cc, line 93 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-YZMBmSKBdpyfg1YL:-L5-YZMBmSKBdpyfg1YM:b-78ue9u) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L93)):*
> ```cpp
>       relative_position[count].first = T(1.);
>     } else {
>       // then input is inside the grid
> ```

BTW _Meta_ this comment and similar above need some more amount of cap /punc.

---

*[math/barycentric.cc, line 97 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-Z7-BFJFwBoXgdfuj:-L5-Z7-BFJFwBoXgdfuk:bbcgukp) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L97)):*
> ```cpp
>       right_index = std::distance(coords.begin(), right_iter);
>       T right_value = *(right_iter);
>       T left_value = *(std::prev(right_iter));
> ```

FYI These could be `const T& foo_value = ...` to avoid a copy.  For `double` it wouldn't matter, though.

---

*[math/barycentric.h, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-b9PI8zwNpzqjtouI:-L5-b9PJcqhYGUIsWLsx:b-7o87n0) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L50)):*
FYI

> for logarithmic lookups

I think, from the API point of view, the rationale for the std::set is that we don't want duplicate values within any one dimension.  The user doesn't really care that we're doing log lookups.

Also, FWIW, my intuition says a vector-of-sortedvector would actually be faster in the implementation (scraping the set into a sorted vector during construction).  The std::binary_search would get us log lookup, and the distance-to-end queries would then become constant-time instead of linear time.  Probably not a big enough difference in practice to merit changing though, until we have performance tests.

---

*[math/barycentric.h, line 100 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-T886LwlazsqDzDF6:-L5-T886LwlazsqDzDF7:b-cbpmmw) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L100)):*
> ```objc
>   ///
>   /// @param mesh_values is a num_outputs by get_num_mesh_points() matrix
>   /// containing the points to interpolate between.
> ```

FYI If I'm understanding this correctly, the order of the columns in `mesh_values` must be consistent with the mesh indices assigned by this class, as exposed by `get_mesh_point`.  That might be worth highlighting in this method's documentation.

---

*[math/test/barycentric_test.cc, line 155 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-cZ1SOJ7dQoPR2rW3:-L5-cZ1SOJ7dQoPR2rW4:b-row1kt) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L155)):*
> ```cpp
>   const int kNumOutputs = 2;
>   MatrixXd mesh(kNumOutputs, bary.get_num_mesh_points());
>   mesh << 1., 2., 3., 4, 5., 6., 7., 8;
> ```

BTW Inconsistent `.` (int vs double) on 4 and 8.

---

*[math/test/barycentric_test.cc, line 155 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-cqSh_CytjKmuHEtq:-L5-cqSh_CytjKmuHEtr:b-r4ih0x) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L155)):*
> ```cpp
>   const int kNumOutputs = 2;
>   MatrixXd mesh(kNumOutputs, bary.get_num_mesh_points());
>   mesh << 1., 2., 3., 4, 5., 6., 7., 8;
> ```

FYI Consider `  // BR` after the 4, so that the code shape matches the matrix shape.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-:-L5-eHzZ613obFmd0sq8:b-te4kzy)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
yeah!  thanks @jwnimmer-tri .  (I've got a few follow-on PRs queued up and ready to go, too, so it will be great to get this in.

---

Review status: 1 of 4 files reviewed at latest revision, 15 unresolved discussions.

---

*[math/barycentric.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4xLPX3hXl5wTRT31Qy:-L5-tR2tM5qHKFR9xY8z:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW punc
</blockquote></details>

Done.

---

*[math/barycentric.cc, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-XClh9JluSU5O3f3K:-L5-tczXNAZMJpFqOfuX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L51)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Should we fail fast if `input` has the wrong size?
</blockquote></details>

Done.

---

*[math/barycentric.cc, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-Y5uyIIHmsQ4GzJZX:-L5-tjzpzDR3wt47Qj2m:b-756zpb) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L56)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI If `T` is `AutoDiffScalar` I believe this will leave is `.value()` initialized to garbage.  We don't need to solve that yet, though.
</blockquote></details>

i think that's ok?  i'm allocating memory, but not depending on any values being written.

---

*[math/barycentric.cc, line 93 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-YZMBmSKBdpyfg1YL:-L5-ty_zR1wyA2fi6ARW:b-6aeqp3) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L93)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW _Meta_ this comment and similar above need some more amount of cap /punc.
</blockquote></details>

i thought someone would say that.  but in my mind, this is the continuation of a sentence/idea, start by the code if().  I almost wrote // ... then input

---

*[math/barycentric.cc, line 97 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-Z7-BFJFwBoXgdfuj:-L5-uIq7KouMdRtjVDy7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L97)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI These could be `const T& foo_value = ...` to avoid a copy.  For `double` it wouldn't matter, though.
</blockquote></details>

Done.

---

*[math/barycentric.h, line 48 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4x7iKe6_hxEYFLj8dC:-L50-MrQkWTEc0xNYCrH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L48)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Missing `#include` for drake_copyable.
</blockquote></details>

Done.

---

*[math/barycentric.h, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-b9PI8zwNpzqjtouI:-L5-ubCSus8v8EznZjQy:btb25iu) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L50)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI

> for logarithmic lookups

I think, from the API point of view, the rationale for the std::set is that we don't want duplicate values within any one dimension.  The user doesn't really care that we're doing log lookups.

Also, FWIW, my intuition says a vector-of-sortedvector would actually be faster in the implementation (scraping the set into a sorted vector during construction).  The std::binary_search would get us log lookup, and the distance-to-end queries would then become constant-time instead of linear time.  Probably not a big enough difference in practice to merit changing though, until we have performance tests.
</blockquote></details>

nice.  i thought about that (but before I fully realized I would be using the std::distance in the middle), and agree with you.  will leave it for now, but have updated the comment.

---

*[math/barycentric.h, line 57 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-OmpW-amSL0T6usmI:-L5-vP9IzJVG-IMb5pot:bd6x1ub) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L57)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI The `input_grid` could also be passed by-value (moved).  Example patch posted at https://gist.github.com/jwnimmer-tri/37b27098f2a28ca34a099bdb68426eda

Advantages of unique_ptr formulation:
- required if `MeshGrid` in the future might be turned into a class that was non-copyable;
- prevents accidentally deep copying (vs forgetting the std::move with by-value).

Advantages of by-value formulation:
- call sites can use initializer lists (i.e., minimal syntax, just the doubles);
- pydrake might be easier to wrap with the plain value.
</blockquote></details>

i was on the fence about it, but am already seeing more advantages of by-value in my next PRs (where I end up using the same mesh for representing multiple functions, and want to take one in but return two separate copies to the user).  I will make the change.  thanks!

---

*[math/barycentric.h, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4x81mMPIJHfOzcZyio:-L5-wvJpfqh44gMy5w_G:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L59)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This comment only binds to the first method.  Suggest either making it `//` or else using Doxygen grouping syntax.
</blockquote></details>

Done.

---

*[math/barycentric.h, line 100 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-T886LwlazsqDzDF6:-L5-xDXotCulU-Fczio5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.h#L100)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI If I'm understanding this correctly, the order of the columns in `mesh_values` must be consistent with the mesh indices assigned by this class, as exposed by `get_mesh_point`.  That might be worth highlighting in this method's documentation.
</blockquote></details>

Done.

---

*[math/test/barycentric_test.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4xLSnHdtvgCt9HrUm7:-L5-xKUdzLKJAmyvsh2n:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW double `at at`
</blockquote></details>

Done.

---

*[math/test/barycentric_test.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4xLVe2u8AR3GuAVlhN:-L5-xgti1Nu8wx1yc00L:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L50)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This duplicateS the comment immediately below; this comment doesn't really appear to describe the case itself at all.
</blockquote></details>

Done.

---

*[math/test/barycentric_test.cc, line 127 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L4xLd2eGmd53nvUfl7L:-L5-xgPv0VBM1DZ3XD77:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L127)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW _Meta_ punc is missing on some of the comments like this (though not all of them are sentences, this and others are).
</blockquote></details>

Done.

---

*[math/test/barycentric_test.cc, line 155 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-cZ1SOJ7dQoPR2rW3:-L5-xyEeZATpH6KR-mBh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L155)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Inconsistent `.` (int vs double) on 4 and 8.
</blockquote></details>

Done.

---

*[math/test/barycentric_test.cc, line 155 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-cqSh_CytjKmuHEtq:-L5-xzD2KvQ-hvJIEl7R:behzjfa) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/test/barycentric_test.cc#L155)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Consider `  // BR` after the 4, so that the code shape matches the matrix shape.
</blockquote></details>

thanks.  yes, clang-format undid it for me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-:-L5-smhg_PofvZ2MqXAw:b-zh1isv)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> as platform.

I'm about halfway through teaching myself the math and checking the implementation (for my own benefit).  (The Munos98 paper didn't help me at all, but I found some other sources.)  In the meantime, I think it would still be good to have a feature review like @hongkai-dai check the documentation and correctness of the arithmetic operations.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[math/barycentric.cc, line 93 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L5-YZMBmSKBdpyfg1YL:-L501T0yY2jQeNAMfu23:b-bfkis) ([raw file](https://github.com/robotlocomotion/drake/blob/d3b0d0794ccc90cc02b9465150bcecc6114a7a50/math/barycentric.cc#L93)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i thought someone would say that.  but in my mind, this is the continuation of a sentence/idea, start by the code if().  I almost wrote // ... then input
</blockquote></details>

I agree (the `...` would be nice but its lack isn't a defect).  I think we still need the punc here though?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-:-L5022hittBmnJZEkX-T:b-b9h0g5)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
First pass done. Also reading the paper to understand the math.

---

Reviewed 1 of 4 files at r1, 2 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[math/barycentric.cc, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L503GxLxAqOAmmMPuSa:-L503GxLxAqOAmmMPuSb:bim5hce) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L42)):*
> ```cpp
>   for (int i = 0; i < get_num_inputs(); i++) {
>     const auto& coords = input_grid_->at(i);
>     int dim_index = index % coords.size();
> ```

`const int dim_index`

---

*[math/barycentric.h, line 62 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L501WtmHYr-SD532t3t:-L501WtmHYr-SD532t3u:b-k83rv4) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.h#L62)):*
> ```objc
> 
>   // Accessor methods.
>   int get_num_inputs() const { return input_grid_->size(); }
> ```

BTW. Should this be called `get_input_size` here? I think we only have one input, but the input is a vector.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-:-L504Vp5vKC-W6ZlBnNX:b-yu2zt4)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[math/barycentric.h, line 107 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L509aBeqkZ9RQ91RDHH:-L508f2wVe-dQRAf7U8b:b4s6t4) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.h#L107)):*
> ```objc
>   /// @param input must be a vector of length get_num_inputs().
>   /// @param output is the interpolated vector of length num_outputs
>   void Eval(const Eigen::Ref<const MatrixX<T>> mesh_values,
> ```

I thought Eigen::Ref should be `const Eigen::Ref<const MatrixX<T>>&` here, that it has `&` in the end. This is explained in https://stackoverflow.com/questions/21132538/correct-usage-of-the-eigenref-class

Same for input below

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
if you find a reference you like better, let me know.  

---

Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[math/barycentric.cc, line 42 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L503GxLxAqOAmmMPuSa:-L50GZ0myjV57YKyfGdk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L42)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`const int dim_index`
</blockquote></details>

Done.

---

*[math/barycentric.h, line 62 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L501WtmHYr-SD532t3t:-L50FoWMWUh6093fQ8bl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.h#L62)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

BTW. Should this be called `get_input_size` here? I think we only have one input, but the input is a vector.
</blockquote></details>

Done.

---

*[math/barycentric.h, line 107 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L509aBeqkZ9RQ91RDHH:-L50G1hTs5yzRgExAUxw:bww7koa) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.h#L107)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I thought Eigen::Ref should be `const Eigen::Ref<const MatrixX<T>>&` here, that it has `&` in the end. This is explained in https://stackoverflow.com/questions/21132538/correct-usage-of-the-eigenref-class

Same for input below
</blockquote></details>

good call.  i forgot.  done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-:-L50GBaIGM8hfNeNQ7hC:b-8tmpyo)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[math/barycentric.cc, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50GhNc-EyObcAgtdm8:-L50GdegHV5IAdHCW9cF:b-15h9rk) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L58)):*
> ```cpp
>   // std::pair of fractional position [0,1] and dimension index (position first,
>   // so that std::pair's default operator< works for us).
>   std::vector<std::pair<T, int>> relative_position(num_interpolants_ - 1);
> ```

Could you explain why `relative_position` is of size `num_interpolants_ -1`, instead of `num_interpolants_`?

---

*[math/barycentric.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50CtcDwEwiSrSixEoc:-L50CtcDwEwiSrSixEod:bpemedz) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L71)):*
> ```cpp
> 
>     // Skip over singleton dimensions.
>     if (coords.size() == 1) continue;
> ```

For singleton dimensions, should we set `has_volume` to false?

---

*[math/barycentric.cc, line 110 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50HENJnQ-RR8ZTxwm5:-L50HENJnQ-RR8ZTxwm6:bh76owy) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L110)):*
> ```cpp
> 
>   // Sort the dimensions by their relative position.
>   std::sort(relative_position.begin(), relative_position.end());
> ```

Why we need to sort the relative position here? This would also make the `weights` sorted. Do we plan to discard the mesh points if their weights are small?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[math/barycentric.cc, line 110 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50HENJnQ-RR8ZTxwm5:-L50MR-4zZVHoW_nmWcP:b2njmka) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L110)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Why we need to sort the relative position here? This would also make the `weights` sorted. Do we plan to discard the mesh points if their weights are small?
</blockquote></details>

OK,  I think I get it. Never mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[math/barycentric.cc, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50GhNc-EyObcAgtdm8:-L50MB7LNdZ5-WGSXkHs:b-f5u47x) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L58)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Could you explain why `relative_position` is of size `num_interpolants_ -1`, instead of `num_interpolants_`?
</blockquote></details>

just added a (long) comment justifying it.  

---

*[math/barycentric.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50CtcDwEwiSrSixEoc:-L50Kj0hRGb5W98crzoD:bgyqi8m) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L71)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

For singleton dimensions, should we set `has_volume` to false?
</blockquote></details>

no.  count  doesn't increment, and there is no relative_position for the singleton dimensions.

---

*[math/barycentric.cc, line 110 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50HENJnQ-RR8ZTxwm5:-L50ML3B8gTRO-iDXD3j:boy65i1) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L110)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

OK,  I think I get it. Never mind.
</blockquote></details>

i've added a longer comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[math/barycentric.cc, line 58 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50GhNc-EyObcAgtdm8:-L50N7jqhgyreqGNcVU_:bcr5s2o) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L58)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

just added a (long) comment justifying it.  
</blockquote></details>

Got it, thanks!

---

*[math/barycentric.cc, line 110 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-L50HENJnQ-RR8ZTxwm5:-L50NWsZJws7UY0GOSuZ:b-jhyysx) ([raw file](https://github.com/robotlocomotion/drake/blob/75378aecd7cd2206754cfb3d8687be4f1f8d6302/math/barycentric.cc#L110)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i've added a longer comment.
</blockquote></details>

Got it. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023#-:-L50NagDdJl_mH_fxoOA:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8023)*
<!-- Sent from Reviewable.io -->
